packages/langium/src/lsp/language-server.ts

```diff
@@ -36,25 +36,53 @@ export function startLanguageServer(services: LangiumServices): void {
                 documentHighlightProvider: {},
                 codeActionProvider: services.lsp.CodeActionProvider ? {} : undefined,
                 // hoverProvider needs to be created for mouse-over events, etc.
-                hoverProvider: false
+                hoverProvider: false,
+                monikerProvider: {}
             }
         };
+
         if (hasWorkspaceFolderCapability) {
             result.capabilities.workspace = {
                 workspaceFolders: {
                     supported: true
                 }
             };
         }
+
+        if (params.capabilities.workspace?.configuration) {
+            try {
+                // experimental
+                if (params.workspaceFolders)
+                    services.index.IndexManager.initializeWorkspace(params.workspaceFolders);
+            } catch (e) {
+                console.error(e);
+            }
+        }
         return result;
     });
 
+    const langiumDocs = services.documents.Documents;
     const documents = services.documents.TextDocuments;
     const documentBuilder = services.documents.DocumentBuilder;
     documents.onDidChangeContent(change => {
-        documentBuilder.build(change.document);
+        console.debug('Documents in service:' + documents.keys().length);
+        langiumDocs.invalidateDocument(change.document.uri);
+        const newDocument = langiumDocs.createOrGetDocument(change.document.uri);
+        if (newDocument.parseResult?.value) {
+            services.index.IndexManager.update(newDocument);
+        }
+        langiumDocs.all.filter(doc => isAffected(doc, change.document.uri)).forEach(
+            doc => {
+                documentBuilder.build(doc);
+                services.index.IndexManager.update(doc);
+            }
+        );
+        documentBuilder.build(newDocument);
```

Comment:
Should this logic be part of `LangiumDocuments`?  Or of `IndexManager`?

Reply:
I was thinking about it as I implemented this parts, but I can't identify any existing service which can/should own this code. This lines are connecting/synching parts between Index and Documents storage; isAffected belongs to that probably too.